zebra-network/src/meta_addr.rs

```diff
@@ -103,51 +410,407 @@ impl PartialOrd for PeerAddrState {
+    ///
+    /// Assmes that missing fields or entries are not clients.
+    pub fn is_direct_client(&self, book_entry: Option<MetaAddr>) -> bool {
+        // Use the latest valid info we have
+        self.into_meta_addr(book_entry)
+            .or(book_entry)
+            .map(|meta_addr| meta_addr.is_direct_client())
+            .unwrap_or(false)
+    }
+
+    /// Apply this change to `old_entry`, returning an updated entry.
+    ///
+    /// [`None`] means "no update".
+    ///
+    /// Ignores out-of-order updates:
+    /// - the same address can arrive from multiple sources,
+    /// - messages are sent by multiple tasks, and
+    /// - multiple tasks read the address book,
+    /// so messaging ordering is not guaranteed.
+    /// (But the address book should eventually be consistent.)
+    pub fn into_meta_addr(self, old_entry: Option<MetaAddr>) -> Option<MetaAddr> {
+        let has_been_attempted = old_entry
+            .map(|old| old.has_been_attempted())
+            .unwrap_or(false);
+
+        let old_last_attempt = old_entry.map(|old| old.get_last_attempt()).flatten();
+        let old_last_success = old_entry.map(|old| old.get_last_success()).flatten();
+        let old_last_failed = old_entry.map(|old| old.get_last_failed()).flatten();
+        let old_untrusted_last_seen = old_entry.map(|old| old.get_untrusted_last_seen()).flatten();
+
+        let old_untrusted_services = old_entry.map(|old| old.get_untrusted_services()).flatten();
+
+        match self {
+            // New seed, not in address book:
+            NewSeed { addr } if old_entry.is_none() => {
+                Some(MetaAddr::new(addr, NeverAttemptedSeed))
+            }
+
+            // New gossiped or alternate, not attempted:
+            // Update state and services, but ignore updates to untrusted last seen
+            NewGossiped {
+                addr,
+                untrusted_services,
+                untrusted_last_seen,
+            } if !has_been_attempted => Some(MetaAddr::new(
+                addr,
+                NeverAttemptedGossiped {
+                    untrusted_services,
+                    // Keep the first time we got
+                    untrusted_last_seen: old_untrusted_last_seen.unwrap_or(untrusted_last_seen),
+                },
+            )),
+            NewAlternate {
+                addr,
+                untrusted_services,
+            } if !has_been_attempted => Some(MetaAddr::new(
+                addr,
+                NeverAttemptedAlternate {
+                    untrusted_services,
+                    // Keep the first time we got
+                    untrusted_last_seen: old_untrusted_last_seen.unwrap_or_else(Utc::now),
+                },
+            )),
+
+            // New entry, but already existing (seed) or attempted (others):
+            // Skip the update entirely
+            NewSeed { .. } | NewGossiped { .. } | NewAlternate { .. } => None,
```

Comment:
NewGossiped and NewAttempted changes are ignored if the peer has been attempted.

So NewGossiped and NewAttempted can replace existing:
* NewGossiped
* NewAttempted
* NewSeed

The services are updated.

Security: the last seen time is kept at the oldest time we have, so that other peers can't keep incrementing the last seen times in our peer set